### DISCUSSION TITLE
Supporting multiple rounding modes

### DISCUSSION THREAD

**Discussion by PaulRBerg** [PaulRBerg/prb-math#186] [Ideas]
## Problem

At the moment, PRBMath supports only one rounding mode: toward zero. This is inconvenient for users who, due to security reasons, require more precise control over the rounding direction.

Small residual amounts (i.e. "dust") can cause functions to revert or lead to potential exploits, e.g. users obtaining something in exchange for nothing. One particular scenario of interest is [ERC-4626](https://eips.ethereum.org/EIPS/eip-4626), which recommends the use of distinct, opposing rounding directions when computing vault shares. In other applications, such as AMMs, the developer may want to round up the price quote to increase their profits.

In this discussion, we will explore the practical ways in which multiple rounding modes could be supported in PRBMath.

## Solution

To start, we need to address the following questions:

1. What new rounding modes should be implemented?
2. Should we modify the current implementations or provide wrappers for existing functions?
3. Where precisely in the code should the new logic be implemented?

Let's examine each question separately.

### New Rounding Modes

The [`Intl.NumberFormat V3`](https://github.com/tc39/proposal-intl-numberformat-v3/tree/842e01#rounding-modes-ecma-402-419) standard specifies nine rounding modes:

1. ceil (toward +∞)
2. floor (toward -∞)
3. expand (away from 0)
4. trunc (toward 0)
5. halfCeil (ties toward +∞)
6. halfFloor (ties toward -∞)
7. halfExpand (ties away from 0; current behavior; default)
8. halfTrunc (ties toward 0)
9. halfEven (ties toward the value with even cardinality)

The first four modes determine the direction of rounding, while the last five employ a rule that rounds to the nearest integer. This requires a tie-breaking rule when the fraction part is exactly 0.5.

For our purposes, we will focus on the direction of rounding and disregard nearest integer rounding (due to the complexity of implementation and gas cost considerations).

Consequently, we have to decide whether to support all possible directions of rounding or just a subset. I would be inclined to start small and iterate, for example, by initially supporting two rounding modes, "expand" and "trunc":

```solidity
enum Rounding {
    Expand, // away from zero
    Trunc // toward zero
}
```
After crossing this first milestone, we can consider the usefulness of the `Ceil` and `Floor` modes.

Which rounding mode would you find useful?

### Wrappers

We will implement rounding modes by introducing wrappers that overload existing functions. This approach is similar to what OpenZeppelin employed in their [Math.sol][oz-math] library.

Here's a quick-and-dirty example for the `log2` function:

```solidity
function log2(UD60x18 x, Rounding rounding) internal pure returns (UD60x18) {
    UD60x18 result = log2(x);
    return result + (rounding == Rounding.Expand && 1 << x < x ? 1 : 0);
}
```

To the extent possible, the existing functions should remain unchanged. We don't want to increase the gas cost for users who are happy with the default rounding mode (`Trunc`).

### Targets

Precision loss occurs only when dividing numbers. Therefore, to identify the targets that require updates to accommodate multiple rounding modes, we need to look for division operations (high-level `div`, assembly `div`, and `/`) and right shifts (`>>`), since right-shifting is equivalent to dividing by a power of two.

- All `mulDiv` functions. It may be necessary to create a new variant of `mulDiv` that ceils the result, i.e. `mulDivCeil`, and rename the current function to `mulDivFloor`
    -`Common.mulDiv` (implicitly `UD60x18.div` and `SD59x18.div`, too)
    - `Common.mulDiv18` (implicitly `UD60x18.mul` and `SD59x18.mul`, too)
    - `Common.mulDivSigned`
- The final calculation in `Common.sqrt`
- The right shift in `UD60x18.avg`
- Most of `UD60x18.avg` and `SD59x18.avg`
- The inline division in `UD60x18.exp` and `SD59x18.exp`
- The inline division in `UD60x18.exp2
- The inline inversion and the division in `SD59x18.exp2`
- All of `UD60x18.inv` and `SD59x18.inv`
- The inline division in `UD60x18.ln` and `SD59x18.ln`
- The inline inversion in `SD59x18.log2`
- The inline division in `UD60x18.log10` and `SD59x18.log10`
- The inline divisions in `UD60x18.pow` when `x` is lower than 1e18

#### Exceptions

There are certain "in-flight" divisions for which adding rounding mode support is not practical or necessary. Here is a comprehensive list of these instances:

- Manual overflow checks, such as those in `SD59x18.sqrt`.
- The integer part calculation in both `UD60x18.log2` and `SD59x18.log2`. These must always round towards zero, as the logarithm computation would not function properly otherwise.
- The calculations for `y` and the repeated halvings (`delta >>= 1`) in `UD60x18.log2` and `SD59x18.log2`. These variables are managed exclusively by the library and should not be influenced by the user. Additionally, incorporating rounding mode support for these calculations would significantly increase the gas cost of using `log2`.
- The repeated halvings (`y >>= 1`) in `UD60x18.powu` and `SD59x18.powu`. The same rationale as for `log2` applies in this case.
- The Newton-Raphson iterations in `Common.sqrt`. For these, the implicit accuracy is good enough since precision converges quadratically, doubling with each iteration.

## Next Steps

In terms of timeline, I cannot give any promises because this project will take a long time to implement. Also, a new major version (V5) may be needed if not all current implementations remain the same.

PRs welcome if anyone wants to take the lead. I can dedicate cycles to reviewing your work.

## Related discussions and resources

- @thedavidmeister's feedback in https://github.com/PaulRBerg/prb-math/discussions/165
- The [Wikipedia article](https://en.wikipedia.org/wiki/Rounding) on rounding
- OpenZeppelin's [`Math.sol`][oz-math] library, which supports three rounding modes (down, up, and zero)
- [Naming rounding modes in `Intl.NumberFormat V3`](https://github.com/tc39/proposal-intl-numberformat-v3/issues/7)
- [Discussion about "banker's rounding" in OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/1712)
- The 4th finding in Trail of Bits' [audit of SAI](https://www.trailofbits.com/documents/sai.pdf)

[oz-math]: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/8d633c/contracts/utils/math/Math.sol

**Comment by thedavidmeister** [comment]
This is a good writeup

Also I generally agree with (and assume that OZ did similar thinking based on their implementation) that:

- Being able to specify rounding is at least sensible and in many cases critical to security
- Default round down behaviour should not cost additional gas
- Generally devs are just looking to be conservative and will prefer overcompensating for a potential exploit over most other considerations (because it can be very difficult to predict edge cases/scenarios due to rounding/dust, even with fuzzing), so "always round up" and "always round down" both fit that mental model depending on context, pragmatically it's probably not worth the complexity, maintenance and gas for anything that doesn't fit that model somehow

Being specific about what can cause the gas cost to spike on the scale of individual mathematical operations, anything that involves jumps is going to ramp things up quickly, so `if` statements and other internal function calls that are difficult or impossible for the optimiser to elide are both things to watch out for.

It would be awesome if Solidity offered a way to have function pointer enums that could dispatch this kind of thing in O(1) but it doesn't afaik so instead we end up with if/else if/else if/else if/... when trying to use an enum to modify behaviour, this can really add up for longer enums found in tight loops (people will want to do basic math like "divide" inside loops, right?). I think this is just an unavoidable consequence of trying to decide something at runtime that we _probably_ know at compile time (seems rare that rounding behaviour would need to be dynamic).

As an implementation detail i'd recommend the enum variant ordering to match the OZ one where compatible, with at least the implied `0` being the default behaviour (round down).

e.g. instead of

```solidity
enum Rounding {
    Expand, // away from zero
    Trunc // toward zero
}
```

do this instead

```solidity
enum Rounding {
    Trunc, // toward zero
    Expand // away from zero
}
```

For just two variants it probably doesn't matter, you're either going to have a single `if`, or more likely take the OZ approach which gives the default logic as directly callable fns with a wrapper that facilitates rounding if someone wants that.

If the _long term_ goal is to go audit all possible rounding modes found in the wild and implement "many" of them, then a more efficient dispatch strategy might be worth looking into at some point.

Code _size_ is another thing to consider (which somewhat contradicts what i just wrote about avoiding jumps 😅 ), for example, i already clocked log10 at `3.683kb` which i think is the worst case in this lib, but if the rounding implementation led to the compiler needing to pull in multiple copies per-rounding-mode then worst case would be 7.2kb+ just for log10+rounding, which is as large as an entire OZ token implementation.

**Comment by thedavidmeister** [comment]
Also I wonder if an "all or nothing" approach and a major version is really needed here, the main current examples of where this is needed is the 4626 spec that literally specifies rounding and AMM/oracle pricing, and all that those both need is `mulDiv`.

There's a very clear 80/20 here imo that could establish a convention/pattern for the lib, and then issues can be backfilled over time to hit the other functions listed above.

Probably next tier of importance would be things like sqrt and inv, and then stuff like avg i find much harder to imagine an _exploit_ due to not being able to round up, although i suppose it's possible.

**Comment by PaulRBerg** [comment]
Potentially relevant: https://github.com/crytic/roundme

**Comment by Keinberger** [comment]
Hi everyone 👋

I’ve gone through this thread thoroughly and would like to contribute a detailed proposal that expands on customizable rounding modes for `PRBMath`. I had previously discussed this idea privately with Paul, and I now want to open it up to the community.

### Problem statement

PRBMath currently rounds consistently towards zero (i.e., truncation). While this behavior is predictable and gas-efficient, it can introduce subtle safety and economic issues, especially in DeFi contexts — as you, @PaulRBerg , already outlined perfectly.

### To illustrate better, let’s take a look at a concrete example

Imagine a protocol that charges interest to users taking out debt. For simplicity, let’s assume the protocol calculates interest based on a constant per-year amount, derived from a system-wide interest rate.

In this context, protocols often prefer to round the interest amount upwards — to avoid undercharging.

But currently, such logic is forced to truncate, meaning the protocol loses small amounts of interest on every operation:

```solidity
uint256 yearlyInterestAmount = ud(principalDebtAmount).mul(ud(yearlyInterestRate)).intoUint256();
```

Ideally, developers should be able to specify the desired rounding direction:

```solidity
uint256 yearlyInterstAmountRounded = ud(principalDebtAmount).mul(ud(yearlyInterestRate), Rounding.Expand).intoUint256();
```

Several prominent libraries already support configurable rounding:

- OpenZeppelin Math (as already mentioned by @PaulRBerg ): Provides an extended `mulDiv` function that accepts a `Rounding` enum to control whether results should round up, down, ceil or floor.
- PercentageMath (Aave): Offers internal rounding logic, which allows for rounding upwards, downwards or also to the nearest integer.
- Solady: Includes precise rounding for arithmetic operations.

> 👉 In my opinion, PRBMath is an extremely precise and developer-friendly library, but this lack of rounding configurability can be a real deal-breaker for developers and protocols choosing between math libraries. I’ve personally encountered this hesitation while building accounting math logic.

### Proposed solution

Here’s how I propose we introduce configurable rounding into PRBMath:

- **Wrappers for existing functions**
    
    I agree with you @PaulRBerg that implementing wrappers is the right direction. The new functionality will be added as overloads or separate methods, meaning no change in gas usage or behavior for existing users. Current users can continue using the standard `mul` and `div` without any disruption.
    
- **Rounding direction specifiable via an enum**
    
    An internal `Rounding` enum is introduced (`Down`, `Up`) to control rounding behavior in an expressive and type-safe manner. This approach matches how OpenZeppelin handles rounding and allows future extensibility. E.g., adding `Ceil`, `Floor`, or even `Nearest`.
    
- Initial focus on `mulDiv`
    
    The new rounding logic will first be implemented for `mulDiv`, and then exposed through wrappers to external endpoints that use `mulDiv`.
    

Here’s a simplified breakdown of how the above changes would translate into code:

`Common.sol`:

```solidity
    /**
     * @notice Enum to specify rounding direction
     * @dev Used for division rounding
     */
    enum Rounding {
        Trunc, // Down
        Expand // Up
    }
    
    /**
     * @notice Wrapper for original mulDiv function
     */
    function mulDiv(uint256 x, uint256 y, uint256 denominator, Rounding rounding) internal pure returns (uint256) {
        uint256 result = mulDiv(x, y, denominator);
        if (rounding == Rounding.Expand && mulmod(x, y, denominator) > 0) {
            result += 1;
        }
        return result;
    }
```

Note: Using `mulmod` similar to OpenZeppelin’s Math.sol library, as it enables us to preserve higher precision for intermediary values exceeding `type(uint256).max`.

`Math.sol`:

```solidity
function mul(UD60x18 x, UD60x18 y, Rounding rounding) pure returns (UD60x18 result) {
    result = wrap(Common.mulDiv(x.unwrap(), y.unwrap(), rounding));
}

function div(UD60x18 x, UD60x18 y, Rounding rounding) pure returns (UD60x18 result) {
    result = wrap(Common.mulDiv(x.unwrap(), uUNIT, y.unwrap(), rounding));
}
```

### Next Steps

I would be happy to:

- Implement the above described logic starting with `mulDiv`, and if accepted, expand it incrementally to external functions that internally use `mulDiv`. To start, I would recommend the core multiplication / division functions, i.e. `mul` & `div`, but happy to expand further.
- Share a full **proof of concept**, complete with tests and documentation updates.
- Ensure minimal to no gas impact for users who don’t opt into the rounding modes.
- Coordinate with you Paul and other contributors for review/feedback on structure and naming conventions.

### Open Questions

1. Should we implement rounding mode logic also for signed math (e.g., `PRBMathSD59x18`)? (Related to the following question)
2. Should we support only Up/Down initially, or add Ceil/Floor from the beginning for future-proofing?
3. @PaulRBerg you shared a list in your original comment of instances where amendments would be needed to support rounding modes. Is this list still accurate?
4. Do we want to add rounding control to **all** external functions that internally use divisions (`mulDiv`)? I personally think it makes sense for most functions, but what about `gm` & `avg`?
5. Are there plans for a major version release (e.g., `v5`) where such changes could live more comfortably alongside others?

**Reply by PaulRBerg** [reply]
Hey @Keinberger thanks so much for the thorough write-up. I have saved this to my to-dos and will review later. Unfortunately, it might take some time, perhaps after EthCC.

**Reply by Keinberger** [reply]
Hey @PaulRBerg , thanks for the quick reply!

If you get a moment, I would really appreciate if you could at least have a quick look at the "Proposed Solution" and "Open Questions" sections. I imagine it won't take more than a few minutes.

That way, I can get started working as soon as possible and you can do a deeper review after EthCC. Does that work for you? :)

**Reply by PaulRBerg** [reply]
## Suggestions

- Let's name the enum value `Expand` and `Trunc` and be consistent with `Intl.NumberFormat V3`. I find `Up`/`Down` confusing.
- Let's implement the rounding wrappers in a new file called `MathRounding.sol` (under `ud60x18`), which will call the functions in `Math.sol`

## Answers

1. No. Signed math is always trickier. Let's have the PoC work only for unsigned math at first.
2. Let's implement only `Expand`/ `Trunc` (`Ceil`/`Floor` can be considered later).
3. Yeah it's still relevant.
4. Yeah all functions should have rounding mode wrappers. But let's skip `gm` and `avg` in the PoC.
5. Yes. This feature would be part of PRBMath v5.

---

P.S. bit of a nitpick but if you will indulge me in it, in future posts could I please ask you to use the `/` feature provided by GitHub to share the code snippets under a [collapsible section](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections)? Doing so would make the comments easier to read through (due to less scrolling).

**Reply by Keinberger** [reply]
Hey @PaulRBerg , thanks for taking a look, Appreciate it!

Regarding your suggestions:

> Let's name the enum value `Expand` and `Trunc` and be consistent with `Intl.NumberFormat V3`. I find `Up`/`Down` confusing.

Agreed, it's definitely best practice to follow existing standards!

> Let's implement the rounding wrappers in a new file called `MathRounding.sol` (under `ud60x18`), which will call the functions in `Math.sol`

Agreed, I think this keeps the code structure more lean!

Also, thanks for answering the questions!

Side note: Will keep the collapsible sections in mind for future posts, thanks for pointing it out!
